File: SEGMENTATION/Document_Segmentation.m
clear all;
clc;
%MAIN FUNCTION

%Image path
path='D:\COLLEGE\1st Sem\IPCV(861)\Project';
name='document4';   %Image name
ext='.jpg';        %Image extension
img_path=strcat(path,'\',name,ext);
rgb = imread(strcat(path,'\',name,ext));
I = rgb2gray(rgb);
Igray=I;


%To find background colour for masking of unwanted regions
histo=imhist(Igray);
[val,ptch]=max(histo);


level=graythresh(I);%uses Otsu’s global thresholding algorithm
BWimage=im2bw(I,level);
%Check points
%imshow(BWimage);
%pause;

%Forming directories
%Makes a folder in the same ddirectors where the image is
folder=strcat(name,'_Segmentation');
cd(path);
[status, msg, msgID] = mkdir(folder);
cd(strcat(path,'\',folder));
folder_path=strcat(path,'\',folder);

[status, msg, msgID] = mkdir('Sentences');
sentence_path=strcat(folder_path,'\','Sentences');
[status, msg, msgID] = mkdir('Words');
word_path=strcat(folder_path,'\','Words');
[status, msg, msgID] = mkdir('Characters');
character_path=strcat(folder_path,'\','Characters');

%Calling sentence segemntation
cd(sentence_path);
n=sentence_segmentation(BWimage,I,Igray,ptch);

%Index to keep a note of the number of images etc...
n=str2num(n);
index=0;

%takes one sentence image at a time
%Sends the image to word segmentation function
%Word segemntation function saves the image is the proper folder
cd(word_path);%directory changed to appropriate folder
disp('starting words');
for i=1:1:n
    name=strcat('sentence',num2str(i),'.jpg');
    img=strcat(sentence_path,'\',name);
    word=imread(img);
    index=word_segmentation1(word,i,index,ptch)
end
%====================

%takes one word image at a time
%Sends the image to char segmentation function
%char segemntation function saves the image is the proper folder
cd(character_path);%directory changed to appropriate folder
disp('starting chars');
disp(index);
n=index;
index=0;

for i=1:1:n
    name=strcat('word',num2str(i),'.jpg');
    img=strcat(word_path,'\',name);
    word=imread(img);
    
    %For programatic approach call this function
    %index=char_segementation_prog(word,i,index);
    
    %watershed approach call this function
    index=char_segementation(word,i,index);
end

File: SEGMENTATION/python/Document_Segmentation.py
import matplotlib.image as img
import cv2
import numpy as np
import os
clear = lambda: os.system('cls')
clear()

def rgb2gray(rgb):

    r, g, b = rgb[:,:,0], rgb[:,:,1], rgb[:,:,2]
    gray = 0.2989 * r + 0.5870 * g + 0.1140 * b

    return gray

def get_hist(image):
    h, w = image.shape[:2]
    histogram = np.zeros([256],np.unit64)
    for i in range(h):
      for j in range(w):
        histogram[image[i][j]] +=1
    return histogram
#MAIN FUNCTION
    
    #Image path
path='C:\Users\Atharva\Desktop\SEM7\Final_Year_project\\'
name='Image_Name'
ext='.jpeg'
img_path = path + name + ext

rgb =img.imread(img_path)
rg = cv2.cvtColor(rgb, cv2.COLOR_RGB2BGR)
I = cv2.cvtColor(rg, cv2.COLOR_BGR2GRAY)
Igray = I

    #To find background colour for masking of unwanted regions
val, ptch = max(get_hist(Igray))
BWimage = cv2.threshold(I, 127, 255, cv2.THRESH_BINARY)
    #Check points
#imshow(I);
#pause;
        #Forming directories
#Makes a folder in the same ddirectors where the image is
folder= name + '_Segmentation'
os.chdir(path)
#cd(path)
status,msg,msgID= os.mkdir(folder)
os.chdir(path + '\' + folder)
folder_path = path +'\' + folder
status,msg,msgID = os.mkdir('Sentences')
sentence_path = folder_path + '\' + 'Sentences'
status,msg,msgID = os.mkdir('Words')
word_path = folder_path + '\' + 'Words'
status,msg,msgID = os.mkdir('Characters')
character_path = folder_path + '\' + 'Characters'

    #Calling sentence segemntation
os.chdir(sentence_path)
n = sentence_segmentation(BWimage,I,Igray,ptch)
#Index to keep a note of the number of images etc...
n= int(n)
index=0

    #takes one sentence image at a time
#Sends the image to word segmentation function
#Word segemntation function saves the image is the proper folder
os.chdir(word_path)
    
print('starting words')
for i in arange(1,n,1).reshape(-1):
    name = 'sentence' + str(i) + '.jpg'
    img = sentence_path + '\' + name
    word = cv2.imread(img)
    index = word_segmentation1(word,i,index,ptch)

    
    #====================
    
    #takes one word image at a time
#Sends the image to char segmentation function
#char segemntation function saves the image is the proper folder
os.chdir(character_path)
print('starting chars')
print(index)
n = index
index=0

for i in arange(1,n,1).reshape(-1):
    name = 'word' + str(i) + '.jpg'
    img = word_path + '\' + name
    word = cv2.imread(img)

        #index=char_segementation_prog(word,i,index);
        #watershed approach call this function
    index = char_segementation(word,i,index)
    